src/s2_cells/mod.rs

```diff
@@ -1,221 +1,263 @@
-        z: -1.0,
-        w: 1.0,
-    },
-    Vector4 {
-        x: -1.0,
-        y: 1.0,
-        z: 1.0,
-        w: 1.0,
-    },
-    Vector4 {
-        x: 1.0,
-        y: -1.0,
-        z: -1.0,
-        w: 1.0,
-    },
-    Vector4 {
-        x: 1.0,
-        y: -1.0,
-        z: 1.0,
-        w: 1.0,
-    },
-    Vector4 {
-        x: 1.0,
-        y: 1.0,
-        z: -1.0,
-        w: 1.0,
-    },
-    Vector4 {
-        x: 1.0,
-        y: 1.0,
-        z: 1.0,
-        w: 1.0,
-    },
-];
+use crate::batch_iterator::{FilteredPointsIterator, PointCloud, PointLocation, PointQuery};
+use crate::data_provider::DataProvider;
+use crate::errors::*;
+use crate::math::{Isometry3, Obb};
+use crate::proto;
+use crate::read_write::{Encoding, NodeIterator};
+use crate::{AttributeDataType, CURRENT_VERSION};
+use cgmath::{Point3, Transform, Vector4};
+use fnv::FnvHashMap;
+use s2::cell::Cell;
+use s2::cellid::CellID;
+use s2::cellunion::CellUnion;
+use s2::point::Point as S2Point;
+use s2::region::Region;
+use std::borrow::Cow;
+use std::collections::HashMap;
+
+pub struct S2Cells {
+    data_provider: Box<dyn DataProvider>,
+    cells: FnvHashMap<CellID, Cell>,
+    meta: S2Meta,
+}
+
+#[derive(Copy, Clone)]
+pub struct S2CellMeta {
+    pub num_points: u64,
+}
+
+impl S2CellMeta {
+    pub fn to_proto(self, cell_id: u64) -> proto::S2Cell {
+        let mut meta = proto::S2Cell::new();
+        meta.set_id(cell_id);
+        meta.set_num_points(self.num_points);
+        meta
+    }
+}
+
+pub struct S2Meta {
+    pub cells: FnvHashMap<CellID, S2CellMeta>,
+    pub attributes: HashMap<String, AttributeDataType>,
+}
+
+impl S2Meta {
+    pub fn to_proto(&self) -> proto::Meta {
+        let cell_protos = self
+            .cells
+            .iter()
+            .map(|(cell_id, cell_meta)| cell_meta.to_proto(cell_id.0))
+            .collect();
+        let mut meta = proto::Meta::new();
+        meta.set_version(CURRENT_VERSION);
+        let mut s2_meta = proto::S2Meta::new();
+        s2_meta.set_cells(::protobuf::RepeatedField::<proto::S2Cell>::from_vec(
+            cell_protos,
+        ));
+        let attributes_meta = self
+            .attributes
+            .iter()
+            .map(|(name, attribute)| {
+                let mut attr_meta = proto::Attribute::new();
+                attr_meta.set_name(name.to_string());
+                attr_meta.set_data_type(attribute.to_proto());
+                attr_meta
+            })
+            .collect();
+        s2_meta.set_attributes(::protobuf::RepeatedField::<proto::Attribute>::from_vec(
+            attributes_meta,
+        ));
+        meta.set_s2(s2_meta);
+        meta
+    }
+
+    pub fn from_proto(meta_proto: proto::Meta) -> Result<Self> {
+        // check if the meta is meant to be for S2 point cloud
+        if !(meta_proto.version == CURRENT_VERSION && meta_proto.has_s2()) {
+            return Err(ErrorKind::InvalidInput("No s2 meta found".to_string()).into());

```

Comment:
How about "Point cloud is not an S2 point cloud"?

Reply:
But this could be confusing if the error actually comes from the version check ... maybe that should have a separate message?